text/0113-secure-script-container.md

```diff
@@ -0,0 +1,183 @@
+status: DRAFTED
+created_at: 2020-07-15
+pr: https://github.com/salesforce/lwc-rfcs/pull/39
+champion: Caridy Patiño, Ted Conn
+---
+
+# Secure Script Container for LWC Apps
+
+## Summary
+
+We need a way for application developers to integrate common third-party libraries into their LWC applications. These libraries, such as Google Analytics, are severely hindered by the shadow DOM which prevents global interaction with any element on the page.
+
+By providing a secure container for scripts to run, see, and interact with all the shadow DOM trees as if it were the light DOM, application developers can continue to run their existing integrations until these libraries support shadow DOM semantics natively.
+
+## Basic example
+
+```html
+<secure-script src="//cdn.optimizely.com/js/12345678.js"></secure-script>
+```
+
+`secure-script` is a custom element which acts the same as the `script` tag, but any code which is evaluated inside this container is able to traverse the entire shadow tree using light DOM semantics. So if the above script runs `document.querySelectorAll('button')`, it will return all button elements regardless if they are in the shadow or not.
+
+```html
+<secure-script>
+    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
+    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
+    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
+    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
+
+    ga('create', 'UA-XXXXX-Y', 'auto');
+    ga('send', 'pageview');
+</secure-script>
+```
+
+Inline scripting would also be supported, such as the above example which uses script injection to fetch its required resources.
+
+## Motivation
+
+Web component applications need the ability to integrate with third party libraries. Because
+web components rely on the shadow DOM for encapsulation, this development paradigm does not work
+with libraries who are expecting to interact with the application globally. Before there was
+component encapsulation via shadow DOM it was possible for library authors to:
+
+- Listen for events and know who fired them
+- Define global styling rules that would get applied on every element
+- Query the entire document for any element on the page
+
+These libraries are typically added as global scripts to the document root rather than deeper 
+integrations into the components themselves. When every component's root element is a shadow
+root, there is no light DOM for them to work with, and their functionality is severely crippled.
+
+Some examples of these libraries are:
+
+- **Analytics tools**: Google Tag Manager / Google Analytics, Pendo
+- **Personalization platforms**: Optimizely, Evergage, Google Optimize
+- **Instrumentation tools**: New Relic, Sentry.io
+
+In addition, developers are finding that this model is not compatible with their **design systems** and **testing frameworks**.
```

Comment:
I would argue that design systems will not fit into this model since they will still be bound to the shadow semantics

Reply:
I agree, this doesn't apply here.